dbms/src/Storages/Page/V3/tests/gtest_page_storage.cpp

```diff
@@ -1491,5 +1493,147 @@ try
+        page_storage = reopenWithConfig(config);
+    }
+
+    const size_t buf_sz = 1024;
+    char c_buff[buf_sz];
+
+    for (size_t i = 0; i < buf_sz; ++i)
+    {
+        c_buff[i] = i % 0xff;
+    }
+
+    PageId page_id0 = 120;
+    {
+        WriteBatch batch;
+        batch.putPage(page_id0, 0, std::make_shared<ReadBufferFromMemory>(c_buff, buf_sz), buf_sz, {});
+        page_storage->write(std::move(batch));
+    }
+
+    // create a snapshot to avoid gc
+    auto snap = page_storage->getSnapshot();
+
+    {
+        WriteBatch batch;
+        batch.delPage(page_id0);
+        page_storage->write(std::move(batch));
+    }
+
+    auto getLogFileNum = [&]() {
+        auto log_files = WALStoreReader::listAllFiles(delegator, Logger::get("PageStorageTest", ""));
+        return log_files.size();
+    };
+
+    // write until there are more than one wal file
+    while (getLogFileNum() <= 1)
+    {
+        WriteBatch batch;
+        PageId page_id1 = 130;
+        batch.putPage(page_id1, 0, std::make_shared<ReadBufferFromMemory>(c_buff, buf_sz), buf_sz, {});
+        page_storage->write(std::move(batch));
+    }
+    ASSERT_ANY_THROW(page_storage->read(page_id0));
+
+    // write an upsert entry into the current writing log file
+    auto done_full_gc = page_storage->gc();
+    EXPECT_TRUE(done_full_gc);
+
+    auto done_snapshot = page_storage->page_directory->tryDumpSnapshot(nullptr, nullptr, /* force */ true);
+    ASSERT_TRUE(done_snapshot);
+
+    {
+        PageStorage::Config config;
+        page_storage = reopenWithConfig(config);
+    }
+
+    ASSERT_ANY_THROW(page_storage->read(page_id0));
+}
+CATCH
+
+TEST_F(PageStorageTest, DumpPageStorageSnapshotWithRefPage)
+try
+{
+    {
+        PageStorage::Config config;
+        config.blob_heavy_gc_valid_rate = 1.0; /// always run full gc
+        config.wal_roll_size = 1 * 1024 * 1024; /// make the wal file more easy to roll
+        config.wal_max_persisted_log_files = 10; /// avoid checkpoint when gc
+        page_storage = reopenWithConfig(config);
+    }
+
+    const size_t buf_sz = 1024;
+    char c_buff[buf_sz];
+
+    for (size_t i = 0; i < buf_sz; ++i)
+    {
+        c_buff[i] = i % 0xff;
+    }
+
+    PageId page_id0 = 120;
+    {
+        WriteBatch batch;
+        batch.putPage(page_id0, 0, std::make_shared<ReadBufferFromMemory>(c_buff, buf_sz), buf_sz, {});
+        page_storage->write(std::move(batch));
+    }
+    PageId page_id1 = 121;
+    {
+        WriteBatch batch;
+        batch.putRefPage(page_id1, page_id0);
+        page_storage->write(std::move(batch));
+    }
+    // create a snapshot to avoid gc
+    auto snap = page_storage->getSnapshot();
+
+    {
+        WriteBatch batch;
+        batch.delPage(page_id0);
+        page_storage->write(std::move(batch));
+    }
+    {
+        WriteBatch batch;
+        batch.delPage(page_id1);
+        page_storage->write(std::move(batch));
+    }
+
```

Comment:
So what if 
1. write page0 
2. ref page 0
3. create a snapshot0
4. del page0
5. write page0 
6. ref page 0
7. create a snapshot1
8. del page0
9. then do the same logic below.
will it still have problem?

Reply:
Yes, the id won't be reused until it is totally deleted.